Use joins for psych sheet events to improve performance - Fixes #731

pingskills opened the pull request:
I used joins to improve the performance in the controller. This seems much, much quicker now. It does rely on user_id being filled out in Preregs but now that all the old competitions have passed I think this is OK.

I also pulled out the looping from the controller as it needs to be done again in the view. I'm not sure if this is the best thing to do from a "rails" perspective.

Also some tests are now failing when using assigns:
`      get :psych_sheet_event, competition_id: competition.id, event_id: "444"
      registrations = assigns(:registrations)`

This seems to return nil. I'd love some help on how to test this.

Let me know your thoughts on the way I went about this and if it's worth finalising and promoting.

larspetrus commented on WcaOnRails/app/views/registrations/psych_sheet_event.html.erb:
```diff
@@ -25,24 +25,38 @@
 
     <tbody>
-      <% @registrations.each do |registration| %>
+      <% current_pos = 0 %>
```
Complex Ruby code should not be in the view.

Not sure what it does exactly, but I'm sure it can be done - and tested - better in the controller.

pingskills replied:
Thanks @larspetrus, I knew that was going to be the answer :) I'll move it back into the controller.

This code loops through the registrations and puts a ranking number ( if 2 people have the same time they get the same number ) against them for this event in the competition. Then the view loops through them in the same manner and prints it out. As the controller and the view were doing basically the same thing twice I thought I'd try just doing it once in the view. Shouldn't be a big deal to move back though to make it easier to maintain and test.

jfly commented:
Sweet stuff! I haven't looked at the code too closely, but I suspect the answer to fixing the broken tests is going to be to do what Lars suggested: keep the position computation in the controller.

It's nice to see some benchmarks when you make a performance change. How many queries were we doing before? How many are we doing now? What does timing look like before and after this change?

pingskills commented:
Thanks @jfly - I'll move the position computation back.

I'll do some research but what's the best way to benchmark a change like this?

In the controller it had this call for each registration - 
`registration.world_rank(@event, @preferred_format.sort_by)`

and in the view it had 4 similar calls per registration:
`          <td class="average"><%= registration.best_solve(@event, :average).clock_format %></td>
          <td class="world-rank-average"><%= registration.world_rank(@event, :average) %></td>

          <td class="single"><%= registration.best_solve(@event, :single).clock_format %></td>
          <td class="world-rank-single"><%= registration.world_rank(@event, :single) %></td>`

So for Euros 2016 that's at least 500 registrations * 5 queries. The new code gets all this information in a single query.

jfly commented:
That sounds significant! In development mode, we print out every sql command, so I'd probably do some trickiness with grep to count them all.

pingskills commented:
OK. I moved the code for computing the position back where it belongs in the controller :)

I then started up the server, loaded the psych_sheet_event for 333 for Euros2016. I stopped the server and moved the log file to new.log. Then I checked out the master branch and did the same thing moving the log to old.log.

Here's the results of the number of select statements:
`[jeff@psdev log]$ grep -i select new.log | wc -l
82
[jeff@psdev log]$ grep -i select old.log | wc -l
15283
[jeff@psdev log]$ `

pingskills commented:
I think the problem with the current test is that `registrations = assigns(:registrations)` now returns nil and I think that is because I use .select

jfly commented:
Holy cow, thats huge! You made sure to empty the log before each of those tests, right?

pingskills commented:
Yes, I emptied the tests before both of the logs. I'll email them to you.

pingskills commented:
I'm not sure why but when I include
`join Countries on Countries.iso2 = users.country_iso2`
in the joins, the assigns(:registrations) in the test returns nil and hence the test fails. If I take that join out, then the assigns works and the test passes? Anybody have any thoughts on this?

pingskills commented:
OK. The problem is that in our test database Countries is not populated so this join fails and no records are returned. I can make it a left join which will mean all of these tests will pass.

pingskills commented:
OK. I've fixed up all the issues so the tests are now successful. I think this is good to go and will make a huge difference to the performance. Can someone please have a close look and let me know if they thinks it's ready.

larspetrus commented:
Can we add countries to the test DB instead?

jonatanklosko commented:
@larspetrus that would eliminate [this](https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/spec/controllers/persons_controller_spec.rb#L18-Lundefined) as well. I was wondering how to do this before and what I came up with is that DatabaseCleaner allows us to avoid cleaning Countries table before/after specs. Have you been doing this before?

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -622,6 +622,66 @@ def continent
+      RanksAverage.worldRank averageRank,
+      ifnull(RanksAverage.best, 0) averageBest,
+      RanksSingle.worldRank singleRank,
+      ifnull(RanksSingle.best, 0) singleBest
```
Please follow snake_case convention =)

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -622,6 +622,66 @@ def continent
+        registration.pos = if registration.tied_previous
+                             prev_registration.pos
+                           else
+                             registration.pos = i + 1
```
This assignment is unnecessary here. Also I'd suggest using ternary here.

jonatanklosko commented on WcaOnRails/app/views/registrations/psych_sheet_event.html.erb:
```diff
@@ -30,19 +30,19 @@
             <%= registration.pos %>
           </td>
-          <td class="name"><%= registration.name %></td>
+          <td class="name"><%= registration.select_name %></td>
```
`registration.name` seems better, is there any reason for using `select_` prefix here?

jonatanklosko replied:
The same question goes for `select_wca_id` and `select_country` below.

pingskills replied:
@jonatanklosko I started with wca_id but the `Registration` model has a method `wca_id` and a method `name`. It seems that in the view when I use `registration.name` it calls the method. That's why I used a different identifier for these.

These methods check if a registration has a user, and if so uses the value from user, otherwise it gets it from Preregs. I think these methods can all go when #275 is done.

larspetrus commented:
Yeah, this should do it:

    config.before(:suite) do
      DatabaseCleaner.clean_with(:truncation, :except => %w[Countries])
    end

Other non dynamic tables that should maybe be permanently populated in the test data are `Continents`, `Events`, `Formats`, 

The 4 `Teams` that are created before each test could maybe also be handled this way.

jonatanklosko commented:
Yeah, that's right. What I actually asked is how would you fill the tables for the first time automatically?

larspetrus commented:
Aha :)

This is how I do it in `spec_helper.rb` in my project:

    config.before(:suite) do
      # -- uncomment to regenerate --
      # Position.delete_all

      if Position.count == 0
        ActiveRecord::Base.transaction do
          BigThought.generate_positions
        end
      end
    end

The test DB gets wiped out in each DB migration, so this gets run then.

pingskills commented:
Our `country` model doesn't reference the database so we may need to make some small adjustments to the spec_helper.rb to get it working.

pingskills commented:
I've updates the code so that the Countries, Continents, Events, Formats, Rounds, and teams tables are not truncated by database_cleaner. The spec_helper.rb will populate these tables if they are empty.

Let me know if you think this is good to push live or if you have some more suggestions. Thanks!

jonatanklosko commented on WcaOnRails/spec/spec_helper.rb:
```diff
@@ -84,4 +84,33 @@
+    if [0] == ActiveRecord::Base.connection.execute("select count(*) from Formats").to_a.first
+      load "#{Rails.root}/db/seeds/formats.seeds.rb"
+    end
+  end
```
Nice!
Nit: that doesn't follow DRY, I'd rewrite this as follows:
```ruby
config.before(:suite) do
  %(Countries Continents Events Rounds Formats teams).each do |table|
    # -- Uncomment to regenerate --
    # ActiveRecord::Base.connection.execute("TRUNCATE #{table};")

    if ActiveRecord::Base.connection.execute("SELECT COUNT(*) FROM #{table}").first == [0]
      load "#{Rails.root}/db/seeds/#{table.underscore}.seeds.rb"
    end
  end
end
```

pingskills replied:
That's much nicer :)

jonatanklosko commented on WcaOnRails/spec/controllers/persons_controller_spec.rb:
```diff
@@ -13,11 +13,6 @@
-      before do
-        # Filled Countries table is necessary here.
-        load "#{Rails.root}/db/seeds/countries.seeds.rb"
-      end
```
Yay!

pingskills commented:
@jonatanklosko I updated the spec_helper.rb code. Nice suggestion. Let me know what you think.

larspetrus commented on WcaOnRails/spec/spec_helper.rb:
```diff
@@ -84,4 +84,15 @@
+      # ActiveRecord::Base.connection.execute("TRUNCATE #{table};")
+
+      if ActiveRecord::Base.connection.execute("SELECT COUNT(*) FROM #{table}").first == [0]
+        load "#{Rails.root}/db/seeds/#{table.underscore}.seeds.rb"
```
Working with strings makes this clumsier than it has to be

    [Countries Continents Events Rounds Formats teams].each do |model_class|
      # -- Uncomment to regenerate --
      # model_class.delete_all

      if model_class.count == 0
        load "#{Rails.root}/db/seeds/#{model_class.name}.seeds.rb"
      end
    end

A few other concerns:

* I don't think using `/db/seeds/` is a great approach here. That's intended to be used with `rake db:seed`, and executing code by using `load` is pretty gross. I'd make a `SeedTestDb` class and call it from here. Not sure where it should live, but it probably doesn't matter much.

* I worry about how this gets updated when the seed data changes. Let's say the `Formats` seed data changes. How does that get propagated to every developer working on their home machine? The `formats` table will not be empty, so it won't trigger on that.

jonatanklosko replied:
Unfortunately we cannot use the code you've presented because the model [`Country`](https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/app/models/country.rb) isn't connected with Countries table (the same goes for all except teams). I remember that Jeremy made the models *virtual* in this way.

larspetrus replied:
Weird. But I believe you.

jonatanklosko replied:
Yeah it's really weird to me as well.
Very good point with seed files changes but I'm not sure how could we deal with this.

larspetrus replied:
I hate any solution including the phrase "...you just have to remember..."!

You could make some kind of checksum on the seed file. But you'd have to store that in the DB. I suppose there could be a 'Seed Checksum" country in the test data... But I don't know how you deal with the checksum itself being in the checksummed file. I suppose it would have to be in a separate class And enforced with a test wherever it's kept.

A bit clunky, and I've never done a file checksum, but it should work, no?

jonatanklosko replied:
I'd move your idea forward and create a file (ignored by git) somewhere inside spec directory called let's say `seeds_hash`. Then in the place we seed the data we concatenate the files contents into string and compute its hash. If it doesn't change then we can skip cleaning the tables. When it changes we write the new checksum to the `seeds_hash` and clean the tables to be reseeded. No need for database mess, what do you think?

larspetrus replied:
Yeah, that's much simpler!

You still need to also check if the DB is empty.

pingskills replied:
@jonatanklosko @larspetrus With the Countries, Continents, Events, Rounds, and Formats this isn't actually a problem. Because they are _virtual_ a migration would need to be written to keep the production database up to date. With the `teams` table though, this could be created through the front end so it is an issue.

We may be over complicating this. it might be simpler to truncate and reload these tables each time. At the moment we are doing one select per table every time to check if they are empty. If we also add the checksum logic it will take a little bit more time. These tables are all small (Continents is the largest with 199 rows, and most are less than 10 rows) so loading them is very quick. This means it won't be much slower to truncate and load each time. As a side note I think we currently insert the 4 teams records before each test so this will be faster than what currently occurs.

I suggest that for this PR I simply truncate and load these tables once each time rspec is called. We can then raise a separate issue if we think this needs to be improved by adding a checksum and only reloading the tables if a seed file has changed. Let me know your thoughts.

larspetrus replied:
When you say "each time", I assume you mean per test suite, not for each test? 

That could make sense. And it's clearly simpler. How long does the data loading take? How long does the whole suite take?

pingskills replied:
Yes I think it's per test suite. Here's the existing code:
```
  config.before(:each) do
    DatabaseCleaner.start
    FactoryGirl.create(:team, friendly_id: 'software', name: 'Software Team', description: 'Does software')
    FactoryGirl.create(:team, friendly_id: 'results', name: 'Results Team', description: 'Posts results')
    FactoryGirl.create(:team, friendly_id: 'wrc', name: 'WRC Team', description: 'Regulations')
    FactoryGirl.create(:team, friendly_id: 'wdc', name: 'WDC Team', description: 'Disciplinary Committee')
  end
```

From memory the whole test suite takes around 90 to 120 seconds to complete. I haven't measured the time the data loads take. I'll check when I get a chance.

pingskills replied:
The loading of data is extremely fast. Here's the results from rails console for loading Countries which is the biggest table:

`=> #<Benchmark::Tms:0x00000006120790 @label="", @real=0.04722644499997841, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.010000000000000009, @total=0.010000000000000009>`

larspetrus replied:
OK, taking 0.04 seconds to set up a 90-120 second test suite is acceptable to me. This also avoids some problems with the test data getting corrupted long term. There is still a possibility of test contamination, where test 23 changes the DB and test 94 has goes wrong asa result. That's probably OK for now.

larspetrus commented on WcaOnRails/spec/support/database_cleaner.rb:
```diff
@@ -1,22 +1,18 @@
 RSpec.configure do |config|
   config.before(:suite) do
-    DatabaseCleaner.clean_with(:truncation)
+    DatabaseCleaner.clean_with(:truncation, except: %w(Countries Continents Events Rounds Formats teams))
```
This is a DRY problem. We should use the same list of classes/tables here as in `spec_helper`, not a copy.

jonatanklosko replied:
I'd suggest moving the `config.before(:suite)` from spec_helper.rb here.

larspetrus commented on WcaOnRails/spec/support/database_cleaner.rb:
```diff
@@ -1,22 +1,18 @@
 
   config.before(:each, js: true) do
-    DatabaseCleaner.strategy = :truncation
+    DatabaseCleaner.strategy = :truncation, {except: %w(Countries Continents Events Rounds Formats teams)}
```
I don't think this is needed, since I don't have it and things work for me :)

I think this only cleans out things created since the beginning of the test, and preexisting data is unaffected.

pingskills replied:
It worked on some tests, but any tests using js it failed. I only added it after the full test suite failed.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -622,6 +622,62 @@ def continent
+                     end
+      registration.pos = nil unless has_competed
+      prev_registration = registration
+    end
```
I felt like it was and still is kinda messy, so I tried something more concise. What do you think about this:
```ruby
rank = "#{preferred_format.sort_by}_rank" # "single_rank" or "average_rank"
prev_registration = nil
registrations.each_with_index do |registration, i|
  registration.tied_previous  = !!(registrtion.send(rank) == prev_registration&.send(rank))
  registration.pos = if registration.tied_previous
                       prev_registration.pos
                     elsif !registration.send(rank)
                       nil # hasn't competed in this even yet
                     else
                       i + 1
                     end
  prev_registration = registration
end
```
Haven't I got too far =) ?

pingskills replied:
@jonatanklosko I don't fully understand what this code does (the !! and registration&. for example) so I prefer my version but since I'm new to ruby if you think this is definitely better let me know and I'll change it.

pingskills replied:
@jonatanklosko For some reason the code you suggested is much slower. It takes around 4 seconds to calculate the 500 positions as opposed to much less than 1 second for the original code. Is this because of the send method? Because of this I might keep it like it is.

jonatanklosko replied:
@pingskills okay I played with this a bit more because I believed this could be really improved.
Firstly: my version works as fast as the original *for me*, I've no idea what's the difference.
Also I've found a way to optimize this even more by stopping computation when we stumble upon a registration that has never competed (as we sort by rank, all subsequent ones hasn't too).

[Here are three methods](http://pastebin.com/4x9R8PCq) I added to Competition model:
- the original (`psych_sheet_event`)
- modified without send (`psych_sheet_event_v2`)
- modified with send (`psych_sheet_event_v3`)

(The part that differs is after assigning `registrations`.)

I've tested them for competition with over 600 accepted registrations for 3x3x3, 20 of them has never competed in it. The code I used for the benchmark is [here](http://pastebin.com/NcKRirV8).
Result (tested in console):
```
original: 0.09891407200999311 (avg of 100)
modified without send: 0.08930202386000019 (avg of 100)
modified with send: 0.0989089906800109 (avg of 100)
```

Of course I've checked that the returned `registrations` are the same for each of these three methods.

I'd be happy with either v2 of v3. I hope that's pretty clean now =)

pingskills replied:
I've gone for option 2. I also had to update one test as the old way used to put a `tied-previous` class on the competitors without results.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -622,6 +622,62 @@ def continent
+                         accepted.
+                         joins(joinsql).
+                         joins(:registration_events).
+                         where("registration_events.event_id=?", event.id).
```
In this particular case I think it would be better to move this two lines to `joinsql` to be consistent.
Then there's no reason to split this into multiple lines.

pingskills replied:
The where clause needs to be separate because the where is going to be automatically added by `.accepted` and to filter for just this competition.

jonatanklosko replied:
I thought about
```ruby
"join registration_events on registration_events.registration_id = Preregs.id and registration_events.event_id = #{event.id}"
```
 instead of where clause. Of course I prefer Rails DSL but joining both with sql and DSL is kinda inconsistent.
(Am I still missing something?)

pingskills replied:
Yeah, that will work. If I was just writing SQL I'd definitely put this in the where clause because it really is a filter not a join condition but if it simplifies things I'm happy to move it into the join condition.

pingskills replied:
@jonatanklosko it doesn't appear to work as it brings back other events besides 333 when this is only specified in the join condition. Here's the SQL that doesn't work:

`SELECT       Preregs.id,       users.name select_name,       users.wca_id select_wca_id,       Preregs.accepted_at,       Countries.name select_country,       registration_events.event_id,       RanksAverage.worldRank average_rank,       ifnull(RanksAverage.best, 0) average_best,       RanksSingle.worldRank single_rank,       ifnull(RanksSingle.best, 0) single_best  FROM `Preregs` join registration_events on registration_events.registration_id = Preregs.id and registration_events.event_id = 333       join users on users.id = Preregs.user_id       join Countries on Countries.iso2 = users.country_iso2       left join RanksSingle on RanksSingle.personId = users.wca_id and RanksSingle.eventId = '333'       left join RanksAverage on RanksAverage.personId = users.wca_id and RanksAverage.eventId = '333' WHERE `Preregs`.`competitionId` = 'Euro2016' AND (`Preregs`.`accepted_at` IS NOT NULL)  ORDER BY -average_rank desc, -single_rank desc, users.name;`

If the condition is in the where clause you get the right result:

`SELECT       Preregs.id,       users.name select_name,       users.wca_id select_wca_id,       Preregs.accepted_at,       Countries.name select_country,       registration_events.event_id,       RanksAverage.worldRank average_rank,       ifnull(RanksAverage.best, 0) average_best,       RanksSingle.worldRank single_rank,       ifnull(RanksSingle.best, 0) single_best  FROM `Preregs` join registration_events on registration_events.registration_id = Preregs.id       join users on users.id = Preregs.user_id       join Countries on Countries.iso2 = users.country_iso2       left join RanksSingle on RanksSingle.personId = users.wca_id and RanksSingle.eventId = '333'       left join RanksAverage on RanksAverage.personId = users.wca_id and RanksAverage.eventId = '333' WHERE `Preregs`.`competitionId` = 'Euro2016' AND (`Preregs`.`accepted_at` IS NOT NULL) and registration_events.event_id = '333'  ORDER BY -average_rank desc, -single_rank desc, users.name;
`

I don't understand why this is happening. If I simplify it for 1 user it looks like it's bringing back any event beginning with 333. I think this is a bug but I'll need to revert to the filter being in the where clause.

`SELECT * FROM `Preregs` join registration_events on registration_events.registration_id = Preregs.id and registration_events.event_id = 333 where competitionId = 'Euro2016' and user_id = 7831`

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -622,6 +622,62 @@ def continent
 
+  def psych_sheet_event(event)
+    preferred_format = event.preferred_formats.first
+    sort_clause = (preferred_format.sort_by == :single) ? "-single_rank desc, -average_rank desc, users.name" : "-average_rank desc, -single_rank desc, users.name"
```
Two thoughts:
- logically would be to define this variable below `joinsql` and `selectsql`
- you can avoid using ternary by doing
```ruby
sort_clause = "-#{preferred_format.sort_by} desc, -#{preferred_format.sort_by_second}_rank desc, users.name"
```

pingskills replied:
Neat. I like that.

jfly commented:
For the record, if we want to get rid of these virtual models in favor of real tables everywhere please go for it. When I created these, I didn't anticipate people wanting to do database joins on these things.

pingskills commented:
I think this is ready now. Let me know your thoughts. The database_cleaner script will truncate and load the reference tables once each time it runs. The actual psych sheet returns really quickly too. :)

pingskills commented:
OK. Nice improvements @jonatanklosko. How are we looking now?

jonatanklosko commented on WcaOnRails/app/views/registrations/psych_sheet_event.html.erb:
```diff
@@ -30,19 +30,19 @@
 
-          <td class="single"><%= registration.best_solve(@event, :single).clock_format %></td>
-          <td class="world-rank-single"><%= registration.world_rank(@event, :single) %></td>
+          <td class="single"><%= SolveTime.new(@event.id, :average, registration.single_best).clock_format %></td>
```
Ah! `:single` should be passed here.

pingskills replied:
Nice catch. This would have been hard to track down. It looks like this would only affect 333fm.

pingskills commented:
Updated. Any other changes?

jonatanklosko commented on WcaOnRails/spec/views/registrations/psych_sheet_event.html.erb_spec.rb:
```diff
@@ -25,22 +27,41 @@
     # Someone who has not competed in 333 before.
-    registrations << FactoryGirl.create(:registration, competition: competition)
-    registrations.last.tied_previous = true
+    registrations << FactoryGirl.create(:registration, :accepted, competition: competition)
```
Hmm I'm a bit confused by the comment above. The registration has user with wca_id (created in the factory), thus it has a person, but this persons hasn't any results. So actually it's someone who has never competed before =) (as above). Maybe we should provide better situation / comment. What do you think?

pingskills replied:
Due to the way the psych_sheet is displayed (we don't reference Persons at all anymore), I don't think we need to distinguish between these 2 differences. I think I should change it to:

```
    # Two users who have not competed before.
    registrations << FactoryGirl.create(:registration, :accepted, competition: competition)
    registrations << FactoryGirl.create(:registration, :accepted, competition: competition)
```

Does that work for you?

jonatanklosko replied:
Good idea to just create two such registrations. Anyway this will create 2 persons (without results but still). Maybe the same as above?
```ruby
2.times do
  registrations << FactoryGirl.create(:registration, :accepted, user: FactoryGirl.create(:user), competition: competition)
end
```

pingskills replied:
Yeah that works well.

jonatanklosko commented:
Great Jeff, this LGTM! Anyway I'd wait for @larspetrus as he was actively participating here =)

larspetrus commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -622,6 +622,56 @@ def continent
+    registrations.each_with_index do |registration, i|
+      if preferred_format.sort_by == :single
+        rank = registration.single_rank
+        prev_rank = prev_registration&.single_rank
```
Nice `&.` trick! Hadn't thought of that.

larspetrus commented on WcaOnRails/spec/support/database_cleaner.rb:
```diff
@@ -1,22 +1,23 @@
+    DatabaseCleaner.clean_with(:truncation, except: reference_tables_to_keep)
+    reference_tables_to_keep.each do |table|
+      ActiveRecord::Base.connection.execute("TRUNCATE #{table};")
+      load "#{Rails.root}/db/seeds/#{table.underscore}.seeds.rb"
```
This still looks quite gross to me. Executing code by `load`, and using `db/seeds` in unintended ways.

I think this should be regular code, perhaps living in `spec/support/PopulateTestDb.rb`.

pingskills replied:
@larspetrus could you please let me know a bit more about how this should work? Does that mean that we'd need to ensure that the seeds files and the PopulateTestDb.rb would need to kept in sync?

We also don't have a method for creating Continents, Countries, Events, Rounds, and Formats so I guess we'd need to use `ActiveRecord::Base.connection.execute`?

larspetrus replied:
I was just thinking we'd move that code to `PopulateTestDb.rb `. Nothing fancy at all. Shouldn't take many minutes.

If it's used somewhere else, either that code could use `PopulateTestDb.rb `, or it's not worth messing with this.

pingskills replied:
Sounds like a good idea. I'll give it a go.

pingskills commented:
@larspetrus I've moved the inserts into a class and then used that class in the seed files and database_cleaner. Let me know what you think.

Rubocop didn't like the load_countries method though :(
Offenses:

spec/support/populate_test_db.rb:2:3: C: Method has too many lines. [201/124]
  def self.load_countries
  ^^^

I don't really want to split this into multiple inserts because that is slower than the way it is currently written. Any suggestions on how to overcome that?

jonatanklosko commented:
@pingskills I'm opposite to moving this SQL to `populate_test_db` file. Those are naturally *seeds* and doesn't belong to test environment. I'd let them stay in the seeds files and do either of this two things:
- use `load <seed file>` from the methods inside PopulateTestDb
- the most simple solution is to do use `%x(bundle exec rake db:seed)` instead of the whole loop, this will also be more automatic, so that any seeds we add in the future will be run as well (assuming we reset them before). I'm not sure how *good* this solution is as I've never called system from code. @larspetrus does this have any drawbacks?

pingskills commented:
@jonatanklosko because we need to truncate the tables, and because we need these tables as exceptions to being truncated, we still need a list of tables.

pingskills commented:
For the moment I reverted the code back to a working set of files that passes all checks. Let me know what you guys decide is best and I'll make the changes.

larspetrus commented:
According to `http://edgeguides.rubyonrails.org/active_record_migrations.html#migrations-and-seed-data`, seed data needs to be in `db/seeds.rb`. If this is literally true, our `db/seeds/countries.seeds.rb` files don't work, which is one reason they seem pointless to me.

But maybe this works anyway, and the documentation is incomplete. Wouldn't be the first time.

I worry that `rake` is slow. If it starts up it's own Rails environment, you don't want that in your test suite.  Haven't tested how it actually performs.

jonatanklosko commented:
@larspetrus we use [seedbank gem](https://github.com/james2m/seedbank) that let us organize our seeds code well. Also it overrides `rake db:seed` so we can happily use it.

pingskills commented:
@larspetrus @jonatanklosko I'm not really sure what the outcome of this was. Would it be OK to merge this up as is, and raise an issue to fix up the seed data? It would be great to get this live because it does currently take a long time to load, and some people have reported timeouts when trying to look at multiple psych sheets at once.

jonatanklosko commented:
I'm okay with the current version =)

larspetrus commented:
Re: seedbank gem: This is why I hate Rails and/or Gems some times. You can never be sure what code actually will do without looking at every single line in the entire project...

In that case, we should of course keep these files.

Not sure what the current state is, but if it needs to be deployed, do it!

pingskills commented:
Deployed. Issue #748 created to discuss the best way for test data to be loaded.
